common/fs/src/tail.rs

```diff
@@ -79,330 +83,396 @@ pub struct Tailer {
-                                        return None;
-                                    }
+                                    debug!("skipping already processed events");
+                                    Metrics::fs().increment_writes();
+                                    return None;
                                 }
+                            }
 
-                                let line = Tailer::handle_event(
-                                    event,
-                                    initial_offsets,
-                                    lookback_config,
-                                    fs.lock().await.deref(),
-                                )
-                                .await;
+                            let line = handle_event(
+                                event,
+                                initial_offsets,
+                                lookback_config,
+                                fs.lock().await.deref(),
+                            )
+                            .await;
 
-                                let line = line.map(|option_val| option_val.map(Ok).right_stream());
+                            let line = line.map(|option_val| option_val.map(Ok).right_stream());
 
-                                if let Some((key, _)) = key_and_previous_event_time {
-                                    let mut event_times = event_times.lock().await;
-                                    let new_event_time = chrono::offset::Utc::now();
-                                    event_times.insert(key, (event_idx, new_event_time));
-                                }
-                                line
+                            if let Some((key, _)) = key_and_previous_event_time {
+                                let mut event_times = event_times.lock().await;
+                                let new_event_time = chrono::offset::Utc::now();
+                                event_times.insert(key, (event_idx, new_event_time));
                             }
+
+                            line
                         }
                     }
                 }
-            })
-            .filter_map(|x| async move { x })
-            .flatten())
+            }
+        })
+        .filter_map(|x| async move { x })
+        .flatten())
+}
+
+impl Tailer {
+    /// Creates new instance of Tailer
+    pub fn new(
+        watched_dirs: Vec<DirPathBuf>,
+        rules: Rules,
+        lookback_config: Lookback,
+        initial_offsets: Option<HashMap<FileId, u64>>,
+    ) -> Self {
+        Self {
+            lookback_config,
+            fs_cache: Arc::new(Mutex::new(FileSystem::new(watched_dirs, rules))),
+            initial_offsets,
+            event_times: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+pin_project! {
+    #[must_use = "streams do nothing unless polled"]
+    pub struct RestartingTailer<P, C, F, S: Stream, Fut> {
```

Comment:
Beginning of the restart stream copy that's specific to passing state in to the closure.